compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3739,12 +3741,34 @@ public void visit(BLangInvocation.BLangResourceAccessInvocation resourceAccessIn
             data.resultType = symTable.semanticError;
         } else {
             BResourceFunction targetResourceFunc = resourceFunctions.get(0);
-            checkExpr(resourceAccessInvocation.resourceAccessPathSegments, targetResourceFunc.resourcePathType, data);
+            checkExpr(resourceAccessInvocation.resourceAccessPathSegments,
+                    getResourcePathType(targetResourceFunc.pathSegmentSymbols), data);
             resourceAccessInvocation.symbol = targetResourceFunc.symbol;
             resourceAccessInvocation.targetResourceFunc = targetResourceFunc;
             checkResourceAccessParamAndReturnType(resourceAccessInvocation, targetResourceFunc, data);
         }
     }
+    
+    private BTupleType getResourcePathType(List<BResourcePathSegmentSymbol> pathSegmentSymbols) {
+        BType restType = null;
+        int pathSegmentCount = pathSegmentSymbols.size();
+        BResourcePathSegmentSymbol lastPathSegmentSym = pathSegmentSymbols.get(pathSegmentSymbols.size() - 1);
+        if (lastPathSegmentSym.name.value.contains("^^")) {
+            restType = lastPathSegmentSym.type;
+            pathSegmentCount--;
+        }
+
+        BTupleType resourcePathType;
+        if (pathSegmentCount > 0 && !lastPathSegmentSym.name.value.contains(".")) {
+            resourcePathType = new BTupleType(pathSegmentSymbols.subList(0, pathSegmentCount).stream()
+                    .map(s -> s.type).collect(Collectors.toList()));
+        } else {
+            resourcePathType = new BTupleType(new ArrayList<>());
+        }
+        
+        resourcePathType.restType = restType;
```

Comment:
If we don't need to assign null to `resourcePathType.restType` we can do `resourcePathType.restType = lastPathSegmentSym.type` at L:3757